### ISSUE TITLE
Add @observes, @unobserves, @on, and @off decorators

### ISSUE THREAD

**Issue by pzuraq** [ember-decorators/ember-decorators#232]
Add the following decorators:

* `@observes`
* `@unobserves`
* `@on`
* `@off`

These will bring the ember decorators project to feature parity with the Ember object-model. `@off` should remove any existing observers or event listeners.

**Comment by allthesignals** [comment]
Would love to assist with this - is anyone else working on it?

**Comment by pzuraq** [comment]
I've looked at it a little bit, `@observes` and `@on` should be easy enough, we can use `Ember.addObserver` and `Ember.addListener` directly. `@off` is trickier, we need to expose some functions in Ember to be able to get the listeners and observers that exist on a function:

* `getObservers`: https://github.com/emberjs/ember.js/blob/2e6edc5d27168bdc3873c5178e1ae3da52ac3cea/packages/ember-utils/lib/super.ts#L47
* `getListeners`: https://github.com/emberjs/ember.js/blob/2e6edc5d27168bdc3873c5178e1ae3da52ac3cea/packages/ember-utils/lib/super.ts#L59

**Comment by machty** [comment]
@pzuraq I'm confused about what `@off` does? Is there some equivalent in the classic EC API? Is it `cancelOn` ?

**Comment by pzuraq** [comment]
It solves the case where an observer or listener is overridden in a subclass:

```js
let Foo = EmberObject.extend({
  foo: observer()
});

let Bar = Foo.extend({
  foo() {

  }
});
```

In this case, the `foo` method should no longer be an observer on Bar. EmberObject accomplishes this when it when Mixins are being collapsed during the `proto` stage. 

Updating this to work with native classes would require some significant work, and given that it is not an incredibly commonly used feature (and that observers and listeners are increasingly recommended against) the `@off` decorator seemed like a better solution. Since observers and event listeners are technically both event listeners, it made sense to use just one decorator for both use cases.

**Comment by lsg-richard** [comment]
Am quite interested in getting `@observes` working as its absence is preventing completing the conversion to ES6 classes in the app I manage. I'd be happy to take a stab at this if someone on the project could provide some initial direction. Also I'm on the community slack as `richard-nz`.

**Comment by ed-asriyan** [comment]
@pzuraq I'm interested in getting `@observes` and `@off` in newer versions of ember-decorators. I've already implemented `@observes` in #267 and I would like to complete the feature by implementing `@off`. I understood its main idea, but I would like to clarify its usage.
 
```javascript
class Foo {
  prop = 1;
  anotherProp = 0;

  @observes('prop', 'anotherProp')
  fooObserver() {
  
  }

  @on('someEvent', 'anotherEvent')
  fooEvent() {

  }
}

class Bar extends Foo {
  @off('prop', 'anotherProp')  // or simply
  fooObserver() {              // @off('prop', 'anotherProp') fooObserver;

  }

  @off('someEvent', 'anotherEvent')  // or simply
  fooEvent() {                       // @off('someEvent', 'anotherEvent') fooObserver;

  }
}

const foo = new Foo();
const bar = new Bar();
set(foo, 'prop', 2); // call fooObserver
set(bar, 'prop', 2); // nothing have called
sendEvent(foo, 'someEvent'); // call fooListener 
sendEvent(bar, 'someEvent'); // nothing have called

```

Is this the right usage example?

**Comment by pzuraq** [comment]
The original intent was for `@off` to not receive any arguments whatsoever, and remove all event listeners and observers on a key. I've talked to some folks about this and we actually think now it would be easier to add two decorators that work like you describe, one for event listeners and one for observers:

```
class Foo {
  prop = 1;
  anotherProp = 0;

  @observes('prop', 'anotherProp')
  fooObserver() {
  
  }

  @on('someEvent', 'anotherEvent')
  fooEvent() {

  }
}

class Bar extends Foo {
  @unobserves('prop', 'anotherProp') fooObserver;
  @off('someEvent', 'anotherEvent') fooObserver;
}
```

There need to be two decorators so we know which type we are applying it too. These would use the reverse APIs of the ones you are using in #267, so `removeObserver` and `removeListener`.
